core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/ParallelShardingMigration.scala

```diff
@@ -0,0 +1,244 @@
+
+  // migrate all shards - except for "_archived". This will include the 4 shards we already migrated, but
+  // those will be quick no-ops because they'll copy zero rows.
+  val shardsToMigrate = (slickDataSource.dataAccess.allShards - "archived").toSeq.sorted
+  val nShards = shardsToMigrate.size
+
+  def migrate() = {
+
+    val tsMigration = System.currentTimeMillis()
+
+    // orphaned-row deletion:
+    // when we migrated the initial 4 shards, we did not delete the migrated rows from
+    // ENTITY_ATTRIBUTE_archived. This saved time (since the deletion is slow), but
+    // it causes problems. When deleting a workspace or deleting individual entities,
+    // these orphaned rows throw foreign-key errors because they still have FKs to ENTITY.
+    //
+    // we do not need to concern ourselves with these orphans if we're doing a full migration
+    // that drops and recreates the _archived table upon completion. This code is here - commented
+    // out - in case we need it at any point.
+    /*
+    logger.warn(s"START deletion of orphan rows.")
+    val tsDeleteStart = System.currentTimeMillis()
+    val deleted = deleteOrphanRows()
+    logger.warn(s"deleted $deleted orphan rows in ${fmt(System.currentTimeMillis()-tsDeleteStart)}ms.")
+     */
+
+    logger.info(s"START migration of $nShards shards.")
+
+    logger.info(s"calculating expected row counts to migrate for each shard (this will take some time) ... ")
+    val expectedCounts = Await.result(Future.traverse(shardsToMigrate) { shardId =>
+      expectedMigrationRowCount(shardId).map(count => (shardId -> count))
+    }.map(_.toMap), Duration.Inf)
+
+    // output rows-to-migrate, sorted by shardid
+    logger.info(s"        SHARD : ROWS TO MIGRATE")
+    expectedCounts.toList.sortBy(_._1).foreach {
+      case (shardId, rowsToMigrate) =>
+        logger.info(s"        $shardId : ${fmt(rowsToMigrate)}")
+    }
+
+    // now, order from least-to-most rows to migrate. In other words,
+    // we migrate the smallest shards first to get them over with.
+    // in the event that anything goes wrong mid-migration, we'll have
+    // gotten at least some of the shards migrated
+    // however, this means that the end of the migration is copying the
+    // largest shards simultaneously in parallel.
+    // val orderedShards = expectedCounts.toList.sortBy(_._2)
+
+    // now, process the shards in *shardId* alphabetical order. This
+    // means the large and small row counts should be pretty evenly distributed,
+    // and thus we are migrating small shards in parallel to
+    // large shards for balanced CPU utilization
+    //
+    // we could also simply use expectedCounts.toList, since that returns a list
+    // in "map order," which is based on the key hashes - also should be evenly distributed.
+    // but, alpha order is easier for us humans to read about in the logs
+    val orderedShards = expectedCounts.toList.sortBy(_._1)
```

Comment:
I think the script will run faster if we reverse the order and start the biggest shards first.  In my testing the slowest shard took over 1.5 hours so if that shard does not get an available thread until way late in the process we'll end up with the job essentially running single-threaded for the last x minutes of execution.  In the grand scheme, it's not a big deal, but if we want to keep the most threads threads busiest for the biggest proportion of program execution, I think the better approach is to start with the biggest shards.

Reply:
yeah - thread count and ordering are the two big places I'm looking to experiment against the real-scale prod clone. Starting with largest first sounds like a good idea, I will definitely evaluate that!